src/mapping/query/function/format_number.rs

```diff
@@ -0,0 +1,254 @@
+use super::prelude::*;
+use bytes::Bytes;
+use rust_decimal::{prelude::FromPrimitive, Decimal};
+
+#[derive(Debug)]
+pub(in crate::mapping) struct FormatNumberFn {
+    query: Box<dyn Function>,
+    scale: Option<Box<dyn Function>>,
+    decimal_separator: Option<Box<dyn Function>>,
+    grouping_separator: Option<Box<dyn Function>>,
+}
+
+impl FormatNumberFn {
+    #[cfg(test)]
+    pub(in crate::mapping) fn new(
+        query: Box<dyn Function>,
+        scale: Option<usize>,
+        decimal_separator: Option<&str>,
+        grouping_separator: Option<&str>,
+    ) -> Self {
+        let scale = scale.map(|v| Box::new(Literal::from(Value::from(v as i64))) as _);
+        let decimal_separator =
+            decimal_separator.map(|v| Box::new(Literal::from(Value::from(v))) as _);
+        let grouping_separator =
+            grouping_separator.map(|v| Box::new(Literal::from(Value::from(v))) as _);
+
+        Self {
+            query,
+            scale,
+            grouping_separator,
+            decimal_separator,
+        }
+    }
+}
+
+impl Function for FormatNumberFn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        let value = required!(ctx, self.query,
+            Value::Integer(v) => Decimal::from_i64(v),
+            Value::Float(v) => Decimal::from_f64(v),
+        )
+        .ok_or("unable to parse number")?;
+
+        let scale = optional!(ctx, self.scale, Value::Integer(v) => v);
+        let grouping_separator = optional!(ctx, self.grouping_separator, Value::Bytes(v) => v);
+        let decimal_separator = optional!(ctx, self.decimal_separator, Value::Bytes(v) => v)
+            .unwrap_or_else(|| Bytes::from("."));
+
+        // Split integral and fractional part of float.
+        let mut parts = value
```

Comment:
I added a `debug_assert`, its length can actually also be 1.

I did have an initial implementation that read it into separate variables, but it made the code more clunky because the second part had to become an `Option`, and the `join` at the end had to be split into multiple steps due to that.